### PR TITLE
 :book: Docs for record events. 

### LINKED ISSUE #2897: [Docs] Document how to record events
### What do you want to happen?

Following up on: https://github.com/kubernetes-sigs/kubebuilder/issues/2684#issuecomment-1227503698

As discussed in the meeting, we think it might be a good idea to add a section about **event recording** to the documentation, similar to [what we had in v1](https://book-v1.book.kubebuilder.io/beyond_basics/creating_events.html).

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by Sajiyah-Salat**
## Descripton

Add the documentation to help out users know how to implement and raise events. 
NOTE: we had this document in v1 docs but it is outdated and we are shaping it accordingly. 

## Motivation

Fixes #2897

### CODE REVIEW

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events:2]: Thank you for your contribution 🥇 .

Following are some comments:

- 1) You need to be able to verify and validate your changes. See that when you push a PR you can visualize how the docs will be within: (Click on Details)

Screenshot 2023-05-01 at 08 59 31

- 2 the doc is missing an entry in the left menu. Can you please add?

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events:9]: We should use the same syntax and CSS that is applied across the project in all docs. 
Can you please check it out? Why are you using `{% panel style="success", title="Getting Events" %}`?

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events:15]: We cannot add those links. The content on them can change. Instead of that, we should add a snippet of the code. 
Also, please look at the other docs and try to follow/adopt the same standards.

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:2]: i dont know how to enter this in left menu. I tried comparing wit quick start for refernece. still dont understand. kindly explain.

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:9]: As mentioned I have just copy pasted the v1 docs and it comes with that. I will surely remove it.

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:15]: https://github.com/Sajiyah-Salat/kubebuilder/blob/master/docs/book/src/migration/legacy/migration_guide_v2tov3.md is this migration guide you are referring to.

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:9]: I think you want me to remove method end method, panel syntax as well.

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:15]: Links are not working. I tried to change it with code snippets. but its not working now. I think some new pr have affect that.

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events:2]: First click on in the details of the CI job for you are able to check the preview of the docs within the changes. Then, look for any left menu title and research for that in the code source so that you will find the file and be able to see how it works.

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events:9]: Check the current docs, see that you can use a search in your IDE or here in the repo to look for the fragment text and find the places so that you know how to do the same.

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events:15]: We should not use the links at all here. We should add examples within code.

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:2]: I am working on this past one hour and still didnt get any left navigation syntax which needed to be added. I tried searching in other pages of reference like generating crds etc. They all just have only one syntax and that is, <aside = note>. I tried searching it on internet and found the 
```
---
tittle: creating-events.md
---
```
method as well. but its not used in other pages. Let me know what I missed now.

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:15]: I know that we dont need to use links. but we need to have code snippets. therefore I want to open the link and get that particular line in the form of code snippets. Therefore opening a link is necessary. wdyt?

File: docs/book/src/reference/creating-events
Comment (Sajiyah-Salat) [reply docs/book/src/reference/creating-events:9]: I have observed the current docs. not more css is used. hardly one or two syntax for notes and code snippets. Therefore I have removed all the syntax.

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events:2]: Following the steps:

#### How to discover how the menu is built

- 1 - Go in the link: https://book.kubebuilder.io/
- 2 - Check any title in the left menu: 

Screenshot 2023-05-04 at 07 28 12

- 3 - Now use the search in your IDE or the repo into github: https://github.com/search?q=repo%3Akubernetes-sigs%2Fkubebuilder%20Available%20Plugins&type=code

- 4 - EURAKA You find all files where you have the same text. So you can discover how, where the menu is done: 

So, the menu is built in the SUMMARY.md file: https://github.com/kubernetes-sigs/kubebuilder/blob/06f81644f0fc2ea8fd07b94c4362535a181ec4f7/docs/book/src/SUMMARY.md?plain=1#L114-L120

Therefore, you need to add a NEW entry for a NEW section, Events. Where? Under references.

#### How to know how to write documentation

- 1 - Now check the docs: https://book.kubebuilder.io/
- 2 - See also the migration docs: https://book.kubebuilder.io/migration/manually_migration_guide_gov3_to_gov4.html
- 3 - See how the steps are provided. Please look at the style used and see how the info is described. 


IMPORTANT: Ensure that you read out the CONTRIBUTING guide: https://github.com/kubernetes-sigs/kubebuilder/blob/master/CONTRIBUTING.md#how-to-contribute-to-docs and check out:

> There are certain writing style guidelines for Kubernetes documentation, check out [style guide](https://kubernetes.io/docs/contribute/style/style-guide/) for more information.  


#### Check out your changes. You must validate if your changes are OK (review it)


- 1 - You need to be able to verify and validate your changes. See that when you push a PR, you can visualize how the docs will be within: (Click on Details)

Screenshot 2023-05-01 at 08 59 31

- 2 - Check out and do your review to see if all IS OK. Then, if so, you ask for reviews. Before that, you can update the title of your PR within (WIP) to make clear that you are still working on it.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Thank you for your contribution, but this PR requires changes.

- Ensure that you check your changes using the preview so that you can auto-verify if all is ok
- You need add a new section in the menu for the new doc (it should be under references)
- You also need to ensure that the content follows up the required styled
- By last, instead of trying to link the samples under testdata/, we must add the code snippet related to the specific examples. 

I hope the detailed steps described in https://github.com/kubernetes-sigs/kubebuilder/pull/3375/files#r1184594890 can help you out.

**Comment by Sajiyah-Salat** [comment]
Thank you so much @camilamacedo86  for these detailed explanation. I am lucky to have a mentor like you.

**Comment by Sajiyah-Salat** [comment]
I know understand how to link page to left menu. All thanks to you. about code snippet, I dont know which code snippet would be used or where to take it. from links one link was working so I just extract it out add it in docs. Let me know which code snippet would you like me to add.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:7]:
```suggestion
 Recent Events for an object can be viewed by running `kubectl describe <resource kind> <resource name>`
```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:8]:
```suggestion

 <aside class="warning">
 
Events should be raised in certain circumstances only. Be aware that it is NOT recommended to emit events for all Operations. If authors raise too many events, it brings bad UX experiences for those consuming the solutions on the cluster, and they will not have attention to what they must be notified of. For further information, see the Kubernetes APIs convention describes when using them [here][Events]. 

</aside>
```

It is a very important information. Therefore, we should:
- a) add on the top to ensure that readers will read this one
- b) use the warning class because we are warning users

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:7]:
```suggestion
 Recent Events for an object can be viewed by running `$ kubectl describe <resource kind> <resource name>`. Also, they can be checked by running `$ kubectl get events`. 
```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:9]:
```suggestion
## How to raise events? 

Following are the steps with examples to help you raise events in your controller's reconciliations. 
```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:11]:
```suggestion
### Passing the Recorder to the Controller

To raise 

```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:12]:
````suggestion
### Passing the EventRecorder to the Controller

Now, we will pass the EventRecorder when we are setting up the controller on the Manager. See that we will change the implementation scaffolded in `cmd/main.go`:

```go
	if err = (&controller.MyKindReconciler{
		Client:   mgr.GetClient(),
		Scheme:   mgr.GetScheme(),
		// Note that we added the following line: 
		Recorder: mgr.GetEventRecorderFor("mykind-controller"),
	}).SetupWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "MyKind")
		os.Exit(1)
	}
```

````

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:13]:
````suggestion
### Granting the required permissions

You must also grant the RBAC rules permissions to allow your project to create Events. Therefore, ensure that you add the [RBAC][rbac-markers] into your controller:

```go
...
//+kubebuilder:rbac:groups=core,resources=events,verbs=create;patch
...

func (r *MyKindReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

And then, run `$ make manifests` to update the rules under `config/rbac/rule.yaml`. 
````

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:33]:
```suggestion
```
We do not need anymore this text.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:46]:
```suggestion
```
We should either remove this part

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:59]:
```suggestion
```
Lets add it on the top.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:106]:
```suggestion
```

We can remove all added here.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:107]:
```suggestion
```

File: docs/book/src/SUMMARY.md
Comment (camilamacedo86) [inline docs/book/src/SUMMARY.md:111]:
```suggestion
  - [Raising Events](./reference/creating-events.md)
```

Can you please add it after `[Using Finalizers](https://github.com/kubernetes-sigs/kubebuilder/blob/da10bf1678133baae9085a96aeffdb9a0dbaa12f/docs/book/src/reference/using-finalizers.md)`

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events.md:13]: We will need to add at the bottom an alias `[rbac-markers]` for the page: https://book.kubebuilder.io/reference/markers/rbac.html

It is a kubebuilder doc, so we should NOT use the above URL; instead, we should use a relative path from this place to the file.

> [rbac-markers]: markers/rbac.md

**Review by camilamacedo86 — COMMENTED** [review]
@Sajiyah-Salat,

I did a full review of this one. Please, feel free to check all suggestions and comments. 

Also, see that I updated the PR description for you to check how we usually do that. We should not add maintainers or issue authors' names. 

The purpose of the description is to be very short content that describes what the PR is about and its motivation.  So that in the future, if we want to check a specific change, we can come back to check it, and also it helps to let the reviewers know what is done and why.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events.md:9]: Let's add the definition with an event within its usage example.
And then, afterwords, provide a detailed example of how to implement it fully. 

It seems more natural. You will first read what it is, how to use it, and its purpose. You only go to the details when/if you need to.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events.md:11]: Lets first change the controller; otherwise, users will check the compiler failing. 
We can also do a generic example and put the imports.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events.md:12]: Now, that we changed the controller, then we can describe how to pass the struct to it.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:14]: You can remove the whole content afterwords. We added all info above already.

**Comment by Sajiyah-Salat** [comment]
I will make suggested changes shortly.

File: docs/book/src/reference/creating-events.md
Comment (varshaprasad96) [inline docs/book/src/reference/creating-events.md:12]:
```suggestion
Be aware that it is **not** recommended to emit Events for all operations. If authors raise too many events, it brings bad UX experiences for those consuming the solutions on the cluster, and they may find it difficult to filter an actionable event from the clutter. For more information, please take a look at the [Kubernetes APIs convention][Events]. 
```

File: docs/book/src/reference/creating-events.md
Comment (varshaprasad96) [inline docs/book/src/reference/creating-events.md:27]: Can we add links for each of the type specifications so that its easier. For example:
event type would be https://github.com/kubernetes/api/blob/6c11c9e4685cc62e4ddc8d4aaa824c46150c9148/core/v1/types.go#L6019-L6024
reason would be https://github.com/kubernetes/api/blob/6c11c9e4685cc62e4ddc8d4aaa824c46150c9148/core/v1/types.go#L6048 
message would be https://github.com/kubernetes/api/blob/6c11c9e4685cc62e4ddc8d4aaa824c46150c9148/core/v1/types.go#L6053

File: docs/book/src/reference/creating-events.md
Comment (varshaprasad96) [inline docs/book/src/reference/creating-events.md:67]: Nit:
```suggestion
c) You can check an example of the event being called in [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303) 
```

Comment (varshaprasad96) [inline docs/book/src/reference/creating-events.md:67]: Nit:
```suggestion
c) You can check an example of the event being called in [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303) 
```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events.md:25]: @Sajiyah-Salat just add the links in the above statements like; 

[(More info)][link-alias] after the explanation I think that would be enough.

Otherwise, all is great for me here and this one has my lgtm

GREAT WORK

Comment (camilamacedo86) [reply docs/book/src/reference/creating-events.md:25]: Can you please squash all commits?
git rebase -i HEAD~<number of commits>
Then, you replace pick with s to squash the commit with the previous one.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:26]: @Sajiyah-Salat see that each spec in the Kubernertes Api Event is described and point out by varsha in the previous suggestion comment.

Please, ensure that you click on in each link to check them out and add the [[more info][link]] for each case scenario.

Also, remember that ALL links should be added as alias and the alias should defined at the bottom.

By last, do not forget to squash all commits and that you addressed all comments and suggestions

File: docs/book/src/reference/creating-events
Comment (camilamacedo86) [reply docs/book/src/reference/creating-events]: See the suggestion made so that you can better check how we can do in this case.

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:23]:
```suggestion
- `eventtype` is this event type, and is either *Normal* or *Warning*. ([More info][Event-Example])
```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:24]:
```suggestion
- `reason` is the reason this event is generated. It should be short and unique with `UpperCamelCase` format. The value could appear in *switch* statements by automation. ([More info][[Reason-Example]])
```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:25]:
```suggestion
- `message` is intended to be consumed by humans. ([More info][Message-Example])
```

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:44]: @Sajiyah-Salat 

It is part of the explanation over how to achieve this goal using the scaffold provided by the Kubebuilder as it is missing parts provided in the suggestions.
The idea is to provide the 3 steps that users must to do for they know how to do that with Kubebuilder projects.

Comment (camilamacedo86) [reply docs/book/src/reference/creating-events.md:44]: See the previous suggestions: https://github.com/kubernetes-sigs/kubebuilder/pull/3375#discussion_r1187045581

File: docs/book/src/reference/creating-events.md
Comment (camilamacedo86) [inline docs/book/src/reference/creating-events.md:45]: @Sajiyah-Salat now see that you would replace the a, b, c for the steps.
It means here we would have the first step as suggested: https://github.com/kubernetes-sigs/kubebuilder/pull/3375#discussion_r1187046166

**Comment by Sajiyah-Salat** [comment]
cc : @camilamacedo86 
Is there anything else you would like me to change.

**Review by camilamacedo86 — COMMENTED** [review]
@Sajiyah-Salat 

Can you please squash the commits so that we can move forward with this one?
After a fast look it seems fine now.

**Comment by Sajiyah-Salat** [comment]
```
git rebase -i HEAD~10
[detached HEAD 92a7859b] Created "creating events" and modified.
 Date: Wed May 17 20:01:15 2023 +0530
 1 file changed, 35 insertions(+), 2 deletions(-)
Auto-merging test/testdata/generate.sh
CONFLICT (content): Merge conflict in test/testdata/generate.sh
error: could not apply 6ee6e8bb... :seedleing: stop to generate all samples for go/v3
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
```
when I run git rebase it gives some conflicting file which have generate something conflicting. We have two option there to choose the current or incoming change or to choose both. What should I choose here?

**Comment by Sajiyah-Salat** [comment]
I ddid messed up the creating-events file and the commits related to it. However we still have the backup file. in another branch. Should I start a new pull request for that.

**Pull request closed without merging**
